include/internal/refcount.h

```diff
@@ -171,4 +171,9 @@ typedef int CRYPTO_REF_COUNT;
 #  define REF_PRINT_COUNT(a, b)
 # endif
 
+
+#define OSSL_UP_REF(c)                                                         \
+int ref = 0;                                                                   \
+return CRYPTO_UP_REF(&c->refcnt, &ref, c->lock) <= 0 ? 0 : (ref > 0)
```

Comment:
I'm undecided whether this auxiliary define is really helpful or whether it just hides information in the locations where it is used. In particular, because it is a function-like macro which can't be used like a function (because of the extra declaration).

Also, IMHO the correct return statement is

```C
return CRYPTO_UP_REF(&c->refcnt, &ref, c->lock) ;
```
and not
```C
return CRYPTO_UP_REF(&c->refcnt, &ref, c->lock) <= 0 ? 0 : (ref > 0) ;
```
for two reasons:

* All variants of CRYPTO_UP_REF() (including the fallback which uses locks), return a boolean value (0 or 1).
* The check `ref > 0` adds extra semantics, which does not belong here.

Reply:
There are quite a few existing ones that have the following pattern:

int DH_up_ref(DH *r)
{
    int i;

    if (CRYPTO_UP_REF(&r->references, &i, r->lock) <= 0)
        return 0;

    REF_PRINT_COUNT("DH", r);
    REF_ASSERT_ISNT(i < 2);
    return ((i > 1) ? 1 : 0);
}

Reply:
I never saw this funny check before.
I wonder if it is correct at all.  Will probably look at it
when I find time.  which may take some time....

Reply:
It seems that on _new(), references is set to 1, and it's debug code to check that you didn't do an up_ref for something where references was 0.

I think we should just replace it with: return CRYPTO_UP_REF(&r->references, &i, r->lock);